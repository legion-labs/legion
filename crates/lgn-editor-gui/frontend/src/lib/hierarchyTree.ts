import { components } from "./path";

export type Entry<Item> = {
  name: string;
  index: number;
  item: Item;
  subEntries: Entry<Item>[] | null;
  // eslint-disable-next-line @typescript-eslint/ban-types
};

// TODO: Improve performance if needed, and stop using recursion
/** A wrapper class around the `Entry<Item>[]` type. */
export class Entries<Item> {
  entries: Entry<Item>[];

  #size: number | null = null;

  /**
   * Build an `Entries` object from any flat arrays of object.
   * Objects must contain a `path` attribute.
   *
   * ## Example
   *
   * ```typescript
   * const entries = Entries.fromArray([
   *   { path: "/foo/bar", value: "hello" },
   *   { path: "/foo/baz", value: "another hello" },
   *   { path: "/foo", value: "another value" },
   *   { path: "baz", value: "a baz value" },
   *   { path: "/foo/bar/baz", value: "another baz value" },
   * ]);
   *
   * const expectedEntries = {
   * };
   *
   * // Given an `assert` and a `deepEqual` function:
   * assert(deepEqual(entries.entries, expectedEntries));
   * ```
   */
  static fromArray<PItem extends { path: string }, AltItem>(
    items: PItem[],
    /** This function is called when an `Item` is not present, typically used for "folders" */
    buildItemFromName: (name: string) => PItem | AltItem
  ): Entries<PItem | AltItem> {
    if (!items.length) {
      return new Entries([]);
    }

    type Ref = {
      [key: string]: Ref;
    } & { subEntries: Entry<PItem | AltItem>[] | null };

    const entriesArray: Entry<PItem>[] = [];
    const ref = { subEntries: entriesArray } as Ref;

    items.forEach((item) => {
      const pathComponents = components(item.path);

      pathComponents.reduce((ref, name, index) => {
        if (!ref[name]) {
          ref[name] = {
            subEntries: index === pathComponents.length - 1 ? null : [],
          } as Ref;

          const entry = {
            name,
            // Dumb index, will be set again properly later
            index: -1,
            item:
              index < pathComponents.length - 1
                ? buildItemFromName(name)
                : item,
            subEntries: ref[name].subEntries,
          };

          if (ref.subEntries) {
            ref.subEntries.push(entry);
          } else {
            ref.subEntries = [entry];
          }
        }

        return ref[name];
      }, ref);
    });

    const entries = new Entries(entriesArray);

    entries.#sort();

    return entries;
  }

  /** Builds an `Entries` object from and array of `Entry` */
  constructor(entries: Entry<Item>[]) {
    this.entries = entries;
  }

  /** Computes the size of the `Entries` */
  get size(): number {
    if (typeof this.#size === "number") {
      return this.#size;
    }

    function count(entries: Entry<Item>[], size = 0): number {
      return entries.reduce(
        (size, entry) =>
          entry.subEntries ? count(entry.subEntries, size + 1) : size + 1,
        size
      );
    }

    this.#size = count(this.entries);

    return this.#size;
  }

  #setIndices() {
    for (const [index, entry] of this) {
      entry.index = index;
    }
  }

  // Could be exposed if needed
  #sort() {
    function sort(entries: Entry<Item>[]): void {
      entries
        .sort((entry1, entry2) => (entry1.name > entry2.name ? 1 : -1))
        .forEach((entry) => {
          if (entry.subEntries?.length) {
            sort(entry.subEntries);
          }
        });
    }

    sort(this.entries);

    this.#setIndices();
  }

  /**
   * Finds an entry in an `Entries` array.
   */
  find(pred: (entry: Entry<Item>) => boolean): Entry<Item> | null {
    for (const [, entry] of this) {
      if (pred(entry)) {
        return entry;
      }
    }

    return null;
  }

  /**
   * Filters `Entries` based on a predicate.
   */
  filter(pred: (entry: Entry<Item>) => boolean): this {
    function filter(entries: Entry<Item>[]): Entry<Item>[] {
      return entries.reduce((acc, entry) => {
        if (!pred(entry)) {
          return acc;
        }

        if (entry.subEntries) {
          return [...acc, { ...entry, subEntries: filter(entry.subEntries) }];
        }

        return [...acc, entry];
      }, [] as Entry<Item>[]);
    }

    this.entries = filter(this.entries);

    // Resets the size so it's computed again if needed
    this.#size = null;

    this.#setIndices();

    return this;
  }

  /**
   * Finds an entry index in an `Entries` array.
   *
   * Unlike `Array.prototype.findIndex`, this method returns `null` if the index is not found, not -1.
   */
  findIndex(pred: (entry: Entry<Item>) => boolean): number | null {
    for (const [index, entry] of this) {
      if (pred(entry)) {
        return index;
      }
    }

    return null;
  }

  /**
   * Takes a whole `Entries` object and a function called for each entry in this object.
   *
   * If the function returns `null` nothing happens,
   * if an item and/or a name is returned, then the entry will be updated.
   */
  update(
    shouldUpdate: (
      entry: Entry<Item>
    ) => Pick<Entry<Item>, "item" | "name"> | null
  ): this {
    function update(entries: Entry<Item>[]): Entry<Item>[] {
      return entries.map((entry) => {
        const updatedEntry = shouldUpdate(entry);

        if (updatedEntry) {
          return {
            ...entry,
            ...updatedEntry,
            name:
              ("name" in updatedEntry && updatedEntry.name.trim()) ||
              entry.name,
          };
        }

        if (!entry.subEntries) {
          return entry;
        }

        return {
          ...entry,
          subEntries: update(entry.subEntries),
        };
      });
    }

    this.entries = update(this.entries);

    this.#sort();

    return this;
  }

  insert<PItem extends { path: string }>(item: PItem): this {
    function insert(
      [part, ...parts]: string[],
      entries: Entry<Item>[]
    ): Entry<Item>[] {
      if (!parts.length) {
        const newEntry: Entry<Item> = {
          index: -1,
          item: item as unknown as Item,
          name: part,
          subEntries: null,
        };

        return [...entries, newEntry];
      }

      const entry = entries.find((entry) => entry.name === part);

      if (!entry) {
        return entries;
      }

      entry.subEntries = insert(parts, entry.subEntries || []);

      return entries;
    }

    this.entries = components(item.path).reduce((acc, part) => {
      acc;

      return acc;
    }, this.entries);

    this.entries = insert(components(item.path), this.entries);

    this.#sort();

    this.#size = null;

    return this;
  }

  /** Get an entry from its index */
  getFromIndex(index: number): Entry<Item> | null {
    for (const [entryIndex, entry] of this) {
      if (entryIndex === index) {
        return entry;
      }
    }

    return null;
  }

  remove(removedEntry: Entry<Item>): this {
    return this.filter((entry) => entry !== removedEntry);
  }

  [Symbol.iterator]() {
    let index = 0;

    function* iter(
      entries: Entry<Item>[]
    ): Generator<[index: number, entry: Entry<Item>]> {
      for (const entry of entries) {
        yield [index++, entry];

        if (entry.subEntries) {
          yield* iter(entry.subEntries);
        }
      }
    }

    return iter(this.entries);
  }
}
