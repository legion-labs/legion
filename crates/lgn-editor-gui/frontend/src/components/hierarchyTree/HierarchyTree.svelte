<script lang="ts">
  import { Entries, updateEntry } from "@/lib/hierarchyTree";
  import clickOutside from "@lgn/frontend/src/actions/clickOutside";
  import { createEventDispatcher } from "svelte";
  import Inner from "./Inner.svelte";

  type Item = $$Generic;

  type $$Slots = {
    itemName: { itemName: string };
  };

  const dispatch = createEventDispatcher<{ select: Item }>();

  export let entries: Entries<Item>;

  export let activeItem: Item | null = null;

  /**
   * This prop function is used to compare 2 items together and must retur
   * `true` if the items are identical.
   *
   * By default `===` is used, so primitives are compared by value and
   * object by reference.
   */
  export let itemsAreIdentical = (item1: Item, item2: Item) => item1 === item2;

  function setActiveItem({ detail: item }: CustomEvent<Item>) {
    activeItem = item;

    dispatch("select", item);
  }

  function setName({
    detail: { item, newName },
  }: CustomEvent<{ newName: string; item: Item }>) {
    entries = updateEntry(entries, (_name, otherItem) =>
      itemsAreIdentical(otherItem, item) ? { name: newName } : null
    );
  }
</script>

<div class="root" use:clickOutside={() => (activeItem = null)}>
  {#each Object.entries(entries) as [name, entry] (name)}
    <Inner
      {entry}
      {activeItem}
      {itemsAreIdentical}
      {name}
      on:select={setActiveItem}
      on:nameChange={setName}
      let:itemName
    >
      <slot name="itemName" slot="itemName" {itemName} />
    </Inner>
  {/each}
</div>

<style lang="postcss">
  .root {
    @apply h-full px-2 overflow-y-auto;
  }
</style>
