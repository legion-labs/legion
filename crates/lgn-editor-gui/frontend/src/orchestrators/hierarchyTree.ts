/**
 * A store orchestrator that contains all the currently displayed items
 * (as a record of hierarchy tree Entries), the entry that's currently being renamed, etc...
 *
 * It could also contain an array of id pointing to the currently expanded entries in the tree.
 */

import type { Writable } from "svelte/store";
import { writable } from "svelte/store";
import type { Entry } from "@/lib/hierarchyTree";
import { Entries } from "@/lib/hierarchyTree";

export type HierarchyTreeOrchestrator<Item extends { path: string }> = {
  currentlyRenameEntry: Writable<Entry<Item> | null>;
  currentEntry: Writable<Entry<Item> | null>;
  entries: Writable<Entries<Item>>;

  load(resources: Item[]): void;
};

export function createHierarchyTreeOrchestrator<Item extends { path: string }>(
  resources: Item[] = []
): HierarchyTreeOrchestrator<Item> {
  return {
    currentlyRenameEntry: writable(null),

    currentEntry: writable(null),

    entries: writable(Entries.fromArray(resources)),

    /** Loads an array of element as hierarchy tree entries in the store */
    load(resources) {
      this.entries.set(Entries.fromArray(resources));
    },
  };
}
