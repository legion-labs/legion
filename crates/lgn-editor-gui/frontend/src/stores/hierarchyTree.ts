/**
 * A store orchestrator that contains all the currently displayed items
 * (as a record of hierarchy tree Entries), the entry that's currently being renamed, etc...
 *
 * It could also contain an array of id pointing to the currently expanded entries in the tree.
 */

import { Entries, Entry } from "@/lib/hierarchyTree";
import { Orchestrator, Writable } from "@lgn/web-client/src/lib/store";

interface IconSetter<Item> {
  (entry: Entry<Item | symbol>): void;
}

export default class<Item extends { path: string }> implements Orchestrator {
  name = "hierarchTree";

  // `symbol` is used only for folder like resources
  // (which will be replaced by proper resources eventually)

  currentlyRenameEntry: Writable<Entry<Item | symbol> | null>;
  entries: Writable<Entries<Item | symbol>>;

  constructor(resources: Item[] = []) {
    this.currentlyRenameEntry = new Writable<Entry<Item | symbol> | null>(null);

    this.entries = new Writable(Entries.fromArray(resources, Symbol));
  }

  /** Loads an array of element as hierarchy tree entries in the store */
  load(resources: Item[], iconSetter: IconSetter<Item> | null) {
    let entries = Entries.fromArray(resources, Symbol);
    if (iconSetter) {
      for (var entry of entries) {
        iconSetter(entry[1]);
      }
    }
    this.entries.set(entries);
  }
}
