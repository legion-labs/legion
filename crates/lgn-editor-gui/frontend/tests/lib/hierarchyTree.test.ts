import { Entries } from "@/lib/hierarchyTree";
import resources from "@/resources/resourcesResponse.json";

describe("updateEntry", () => {
  test("updates no entry names when the provided function always returns null in the Entries", () => {
    const entries = Entries.fromArray(resources, Symbol);

    expect(entries.update(() => null)).toEqual(entries);
  });

  test("updates no entry names when the provided function always return an empty string in the Entries", () => {
    const entries = Entries.fromArray(resources, Symbol);

    expect(
      entries.update((entry) => ({
        ...entry,
        name: "",
      }))
    ).toEqual(entries);
  });

  test('updates 1 entry name "leaf" when the provided function returns a non empty string for a "leaf" entry', () => {
    const entries = Entries.fromArray(resources, Symbol);

    expect(
      entries.update((entry) =>
        entry.name === "cube_group.ent"
          ? { ...entry, name: "new_cube_group.ent" }
          : null
      )
    ).toMatchSnapshot();
  });

  test('updates 1 entry name "node" when the provided function returns a non empty string for a "node" entry', () => {
    const entries = Entries.fromArray(resources, Symbol);

    expect(
      entries.update((entry) =>
        entry.name === "world" ? { ...entry, name: "monde" } : null
      )
    ).toMatchSnapshot();
  });
});

describe("fromArray", () => {
  test("transforms a `ResourceDescription` array into a hierarchical tree", () => {
    expect(Entries.fromArray(resources, Symbol)).toMatchSnapshot();
  });
});
