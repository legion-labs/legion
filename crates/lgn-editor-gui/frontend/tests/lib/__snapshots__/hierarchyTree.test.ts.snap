// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromArray transforms a \`ResourceDescription\` array into a hierarchical tree 1`] = `
Entries {
  "entries": Array [
    Object {
      "index": 0,
      "item": Symbol(entity),
      "name": "entity",
      "subEntries": Array [
        Object {
          "index": 1,
          "item": Object {
            "id": "(09cde0390abe0694,db051b98-6ff5-4bac-bea8-50b5a13c3f1b)",
            "path": "/entity/DebugCube0",
            "version": 1,
          },
          "name": "DebugCube0",
          "subEntries": null,
        },
        Object {
          "index": 2,
          "item": Object {
            "id": "(09cde0390abe0694,202e3aa6-f158-4c77-890b-3f59b183b6bd)",
            "path": "/entity/DebugCube1",
            "version": 1,
          },
          "name": "DebugCube1",
          "subEntries": null,
        },
        Object {
          "index": 3,
          "item": Object {
            "id": "(09cde0390abe0694,7483c534-fe2a-4f16-b655-e9afe39a93ba)",
            "path": "/entity/DebugCube2",
            "version": 1,
          },
          "name": "DebugCube2",
          "subEntries": null,
        },
        Object {
          "index": 4,
          "item": Object {
            "id": "(0bac9261332b6ce8,d8fe06a0-1317-46f5-902b-266b0eae6fa8)",
            "path": "/entity/TEST_ENTITY_NAME.dc",
            "version": 1,
          },
          "name": "TEST_ENTITY_NAME.dc",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 5,
      "item": Symbol(image),
      "name": "image",
      "subEntries": Array [
        Object {
          "index": 6,
          "item": Object {
            "id": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)",
            "path": "/image/ground.psd",
            "version": 1,
          },
          "name": "ground.psd",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 7,
      "item": Symbol(prefab),
      "name": "prefab",
      "subEntries": Array [
        Object {
          "index": 8,
          "item": Symbol(props),
          "name": "props",
          "subEntries": Array [
            Object {
              "index": 9,
              "item": Object {
                "id": "(1c0ff9e497b0740f,270c28b5-6368-4a21-8ef1-a7d03f85abb3)",
                "path": "/prefab/props/cube.ent",
                "version": 1,
              },
              "name": "cube.ent",
              "subEntries": null,
            },
            Object {
              "index": 10,
              "item": Object {
                "id": "(1c0ff9e497b0740f,45d44b64-b97c-486d-a0d7-151974c28263)",
                "path": "/prefab/props/cube_group.ent",
                "version": 1,
              },
              "name": "cube_group.ent",
              "subEntries": null,
            },
            Object {
              "index": 11,
              "item": Object {
                "id": "(91dc60590cba171e,7d0e930c-eb69-45c4-a695-fcb787f3700f)",
                "path": "/prefab/props/cube_group_cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 12,
              "item": Object {
                "id": "(91dc60590cba171e,c6db12c5-47a6-407d-94da-b6853f36eb8a)",
                "path": "/prefab/props/cube_group_cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_2.ent.ins",
              "subEntries": null,
            },
          ],
        },
      ],
    },
    Object {
      "index": 13,
      "item": Symbol(world),
      "name": "world",
      "subEntries": Array [
        Object {
          "index": 14,
          "item": Symbol(sample_1),
          "name": "sample_1",
          "subEntries": Array [
            Object {
              "index": 15,
              "item": Object {
                "id": "(91dc60590cba171e,7d083d64-cbf9-4786-ae8b-44924b3e2035)",
                "path": "/world/sample_1/cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 16,
              "item": Object {
                "id": "(91dc60590cba171e,ee30f104-3ef5-46e7-af16-4f085bd6f8e3)",
                "path": "/world/sample_1/cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_2.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 17,
              "item": Object {
                "id": "(91dc60590cba171e,4f595c8b-3b9a-41d6-9cf6-a7557dd2e8ee)",
                "path": "/world/sample_1/cube_3.ent.ins",
                "version": 1,
              },
              "name": "cube_3.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 18,
              "item": Object {
                "id": "(91dc60590cba171e,181f9358-000f-47db-9583-7c9448c3dba8)",
                "path": "/world/sample_1/cube_group_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 19,
              "item": Object {
                "id": "(1c0ff9e497b0740f,e07c7027-a823-47e5-b0dd-8871e27adfce)",
                "path": "/world/sample_1/ground.ent",
                "version": 1,
              },
              "name": "ground.ent",
              "subEntries": null,
            },
            Object {
              "index": 20,
              "item": Object {
                "id": "(07dd9f5d1793ed64,48909c46-ad4f-4d6b-a522-2e16e81ba082)",
                "path": "/world/sample_1/ground.mat",
                "version": 1,
              },
              "name": "ground.mat",
              "subEntries": null,
            },
            Object {
              "index": 21,
              "item": Object {
                "id": "(80b97dd7e2e9edca,7948dc8c-8bc6-4aaf-b3cf-0a21633edc44)",
                "path": "/world/sample_1/ground.mesh",
                "version": 1,
              },
              "name": "ground.mesh",
              "subEntries": null,
            },
          ],
        },
        Object {
          "index": 22,
          "item": Object {
            "id": "(1c0ff9e497b0740f,5d6c8521-ef7f-4402-8fac-01c5c4f53329)",
            "path": "/world/sample_1.ent",
            "version": 1,
          },
          "name": "sample_1.ent",
          "subEntries": null,
        },
      ],
    },
  ],
}
`;

exports[`updateEntry updates 1 entry name "leaf" when the provided function returns a non empty string for a "leaf" entry 1`] = `
Entries {
  "entries": Array [
    Object {
      "index": 0,
      "item": Symbol(entity),
      "name": "entity",
      "subEntries": Array [
        Object {
          "index": 1,
          "item": Object {
            "id": "(09cde0390abe0694,db051b98-6ff5-4bac-bea8-50b5a13c3f1b)",
            "path": "/entity/DebugCube0",
            "version": 1,
          },
          "name": "DebugCube0",
          "subEntries": null,
        },
        Object {
          "index": 2,
          "item": Object {
            "id": "(09cde0390abe0694,202e3aa6-f158-4c77-890b-3f59b183b6bd)",
            "path": "/entity/DebugCube1",
            "version": 1,
          },
          "name": "DebugCube1",
          "subEntries": null,
        },
        Object {
          "index": 3,
          "item": Object {
            "id": "(09cde0390abe0694,7483c534-fe2a-4f16-b655-e9afe39a93ba)",
            "path": "/entity/DebugCube2",
            "version": 1,
          },
          "name": "DebugCube2",
          "subEntries": null,
        },
        Object {
          "index": 4,
          "item": Object {
            "id": "(0bac9261332b6ce8,d8fe06a0-1317-46f5-902b-266b0eae6fa8)",
            "path": "/entity/TEST_ENTITY_NAME.dc",
            "version": 1,
          },
          "name": "TEST_ENTITY_NAME.dc",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 5,
      "item": Symbol(image),
      "name": "image",
      "subEntries": Array [
        Object {
          "index": 6,
          "item": Object {
            "id": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)",
            "path": "/image/ground.psd",
            "version": 1,
          },
          "name": "ground.psd",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 7,
      "item": Symbol(prefab),
      "name": "prefab",
      "subEntries": Array [
        Object {
          "index": 8,
          "item": Symbol(props),
          "name": "props",
          "subEntries": Array [
            Object {
              "index": 9,
              "item": Object {
                "id": "(1c0ff9e497b0740f,270c28b5-6368-4a21-8ef1-a7d03f85abb3)",
                "path": "/prefab/props/cube.ent",
                "version": 1,
              },
              "name": "cube.ent",
              "subEntries": null,
            },
            Object {
              "index": 10,
              "item": Object {
                "id": "(91dc60590cba171e,7d0e930c-eb69-45c4-a695-fcb787f3700f)",
                "path": "/prefab/props/cube_group_cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 11,
              "item": Object {
                "id": "(91dc60590cba171e,c6db12c5-47a6-407d-94da-b6853f36eb8a)",
                "path": "/prefab/props/cube_group_cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_2.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 12,
              "item": Object {
                "id": "(1c0ff9e497b0740f,45d44b64-b97c-486d-a0d7-151974c28263)",
                "path": "/prefab/props/cube_group.ent",
                "version": 1,
              },
              "name": "new_cube_group.ent",
              "subEntries": null,
            },
          ],
        },
      ],
    },
    Object {
      "index": 13,
      "item": Symbol(world),
      "name": "world",
      "subEntries": Array [
        Object {
          "index": 14,
          "item": Symbol(sample_1),
          "name": "sample_1",
          "subEntries": Array [
            Object {
              "index": 15,
              "item": Object {
                "id": "(91dc60590cba171e,7d083d64-cbf9-4786-ae8b-44924b3e2035)",
                "path": "/world/sample_1/cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 16,
              "item": Object {
                "id": "(91dc60590cba171e,ee30f104-3ef5-46e7-af16-4f085bd6f8e3)",
                "path": "/world/sample_1/cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_2.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 17,
              "item": Object {
                "id": "(91dc60590cba171e,4f595c8b-3b9a-41d6-9cf6-a7557dd2e8ee)",
                "path": "/world/sample_1/cube_3.ent.ins",
                "version": 1,
              },
              "name": "cube_3.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 18,
              "item": Object {
                "id": "(91dc60590cba171e,181f9358-000f-47db-9583-7c9448c3dba8)",
                "path": "/world/sample_1/cube_group_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 19,
              "item": Object {
                "id": "(1c0ff9e497b0740f,e07c7027-a823-47e5-b0dd-8871e27adfce)",
                "path": "/world/sample_1/ground.ent",
                "version": 1,
              },
              "name": "ground.ent",
              "subEntries": null,
            },
            Object {
              "index": 20,
              "item": Object {
                "id": "(07dd9f5d1793ed64,48909c46-ad4f-4d6b-a522-2e16e81ba082)",
                "path": "/world/sample_1/ground.mat",
                "version": 1,
              },
              "name": "ground.mat",
              "subEntries": null,
            },
            Object {
              "index": 21,
              "item": Object {
                "id": "(80b97dd7e2e9edca,7948dc8c-8bc6-4aaf-b3cf-0a21633edc44)",
                "path": "/world/sample_1/ground.mesh",
                "version": 1,
              },
              "name": "ground.mesh",
              "subEntries": null,
            },
          ],
        },
        Object {
          "index": 22,
          "item": Object {
            "id": "(1c0ff9e497b0740f,5d6c8521-ef7f-4402-8fac-01c5c4f53329)",
            "path": "/world/sample_1.ent",
            "version": 1,
          },
          "name": "sample_1.ent",
          "subEntries": null,
        },
      ],
    },
  ],
}
`;

exports[`updateEntry updates 1 entry name "node" when the provided function returns a non empty string for a "node" entry 1`] = `
Entries {
  "entries": Array [
    Object {
      "index": 0,
      "item": Symbol(entity),
      "name": "entity",
      "subEntries": Array [
        Object {
          "index": 1,
          "item": Object {
            "id": "(09cde0390abe0694,db051b98-6ff5-4bac-bea8-50b5a13c3f1b)",
            "path": "/entity/DebugCube0",
            "version": 1,
          },
          "name": "DebugCube0",
          "subEntries": null,
        },
        Object {
          "index": 2,
          "item": Object {
            "id": "(09cde0390abe0694,202e3aa6-f158-4c77-890b-3f59b183b6bd)",
            "path": "/entity/DebugCube1",
            "version": 1,
          },
          "name": "DebugCube1",
          "subEntries": null,
        },
        Object {
          "index": 3,
          "item": Object {
            "id": "(09cde0390abe0694,7483c534-fe2a-4f16-b655-e9afe39a93ba)",
            "path": "/entity/DebugCube2",
            "version": 1,
          },
          "name": "DebugCube2",
          "subEntries": null,
        },
        Object {
          "index": 4,
          "item": Object {
            "id": "(0bac9261332b6ce8,d8fe06a0-1317-46f5-902b-266b0eae6fa8)",
            "path": "/entity/TEST_ENTITY_NAME.dc",
            "version": 1,
          },
          "name": "TEST_ENTITY_NAME.dc",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 5,
      "item": Symbol(image),
      "name": "image",
      "subEntries": Array [
        Object {
          "index": 6,
          "item": Object {
            "id": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)",
            "path": "/image/ground.psd",
            "version": 1,
          },
          "name": "ground.psd",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 7,
      "item": Symbol(world),
      "name": "monde",
      "subEntries": Array [
        Object {
          "index": 8,
          "item": Symbol(sample_1),
          "name": "sample_1",
          "subEntries": Array [
            Object {
              "index": 9,
              "item": Object {
                "id": "(91dc60590cba171e,7d083d64-cbf9-4786-ae8b-44924b3e2035)",
                "path": "/world/sample_1/cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 10,
              "item": Object {
                "id": "(91dc60590cba171e,ee30f104-3ef5-46e7-af16-4f085bd6f8e3)",
                "path": "/world/sample_1/cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_2.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 11,
              "item": Object {
                "id": "(91dc60590cba171e,4f595c8b-3b9a-41d6-9cf6-a7557dd2e8ee)",
                "path": "/world/sample_1/cube_3.ent.ins",
                "version": 1,
              },
              "name": "cube_3.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 12,
              "item": Object {
                "id": "(91dc60590cba171e,181f9358-000f-47db-9583-7c9448c3dba8)",
                "path": "/world/sample_1/cube_group_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 13,
              "item": Object {
                "id": "(1c0ff9e497b0740f,e07c7027-a823-47e5-b0dd-8871e27adfce)",
                "path": "/world/sample_1/ground.ent",
                "version": 1,
              },
              "name": "ground.ent",
              "subEntries": null,
            },
            Object {
              "index": 14,
              "item": Object {
                "id": "(07dd9f5d1793ed64,48909c46-ad4f-4d6b-a522-2e16e81ba082)",
                "path": "/world/sample_1/ground.mat",
                "version": 1,
              },
              "name": "ground.mat",
              "subEntries": null,
            },
            Object {
              "index": 15,
              "item": Object {
                "id": "(80b97dd7e2e9edca,7948dc8c-8bc6-4aaf-b3cf-0a21633edc44)",
                "path": "/world/sample_1/ground.mesh",
                "version": 1,
              },
              "name": "ground.mesh",
              "subEntries": null,
            },
          ],
        },
        Object {
          "index": 16,
          "item": Object {
            "id": "(1c0ff9e497b0740f,5d6c8521-ef7f-4402-8fac-01c5c4f53329)",
            "path": "/world/sample_1.ent",
            "version": 1,
          },
          "name": "sample_1.ent",
          "subEntries": null,
        },
      ],
    },
    Object {
      "index": 17,
      "item": Symbol(prefab),
      "name": "prefab",
      "subEntries": Array [
        Object {
          "index": 18,
          "item": Symbol(props),
          "name": "props",
          "subEntries": Array [
            Object {
              "index": 19,
              "item": Object {
                "id": "(1c0ff9e497b0740f,270c28b5-6368-4a21-8ef1-a7d03f85abb3)",
                "path": "/prefab/props/cube.ent",
                "version": 1,
              },
              "name": "cube.ent",
              "subEntries": null,
            },
            Object {
              "index": 20,
              "item": Object {
                "id": "(1c0ff9e497b0740f,45d44b64-b97c-486d-a0d7-151974c28263)",
                "path": "/prefab/props/cube_group.ent",
                "version": 1,
              },
              "name": "cube_group.ent",
              "subEntries": null,
            },
            Object {
              "index": 21,
              "item": Object {
                "id": "(91dc60590cba171e,7d0e930c-eb69-45c4-a695-fcb787f3700f)",
                "path": "/prefab/props/cube_group_cube_1.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_1.ent.ins",
              "subEntries": null,
            },
            Object {
              "index": 22,
              "item": Object {
                "id": "(91dc60590cba171e,c6db12c5-47a6-407d-94da-b6853f36eb8a)",
                "path": "/prefab/props/cube_group_cube_2.ent.ins",
                "version": 1,
              },
              "name": "cube_group_cube_2.ent.ins",
              "subEntries": null,
            },
          ],
        },
      ],
    },
  ],
}
`;
